compiler/rustc_codegen_ssa/src/traits/consts.rs

```diff
@@ -39,7 +51,102 @@ pub trait ConstCodegenMethods: BackendTypes {
 
     fn const_data_from_alloc(&self, alloc: ConstAllocation<'_>) -> Self::Value;
 
-    fn scalar_to_backend(&self, cv: Scalar, layout: abi::Scalar, llty: Self::Type) -> Self::Value;
+    fn const_bitcast(&self, val: Self::Value, ty: Self::Type) -> Self::Value;
+    fn const_pointercast(&self, val: Self::Value, ty: Self::Type) -> Self::Value;
+    fn const_int_to_ptr(&self, val: Self::Value, ty: Self::Type) -> Self::Value;
+    fn const_ptr_to_int(&self, val: Self::Value, ty: Self::Type) -> Self::Value;
+    /// Create a global constant.
+    ///
+    /// The returned global variable is a pointer in the default address space for globals.
+    fn static_addr_of_impl(&self, cv: Self::Value, align: Align, kind: Option<&str>)
+    -> Self::Value;
+
+    /// Same as `static_addr_of_impl`, but does not mark the static as immutable
+    fn static_addr_of_mut(&self, cv: Self::Value, align: Align, kind: Option<&str>) -> Self::Value;
+
+    fn scalar_to_backend(&self, cv: Scalar, layout: abi::Scalar, llty: Self::Type) -> Self::Value {
+        let bitsize = if layout.is_bool() { 1 } else { layout.size(self).bits() };
+        match cv {
+            Scalar::Int(int) => {
+                let data = int.to_bits(layout.size(self));
+                let llval = self.const_uint_big(self.type_ix(bitsize), data);
+                if matches!(layout.primitive(), Primitive::Pointer(_)) {
+                    self.const_int_to_ptr(llval, llty)
+                } else {
+                    self.const_bitcast(llval, llty)
+                }
+            }
```

Comment:
This is very likely to break cg_gcc since the bitcast doesn't work in const context (see comment in the original implementation).
I can check if I could make it work in const context if needed.

Nice refactor, btw!

Reply:
can we cheat and do the `bitsize > 1 && ty.is_integral() && bytesize as u32 == ty.get_size()` check within the cg_gcc `const_bitcast` impl?

Reply:
I'll have to thing about whether this could break other things: if that ever gets called with a value of a struct type, for instance, I assume this would break.
Would you have access to `bytesize`?

Reply:
we can very readily provide it, we're in control of this method and llvm will just ignore it

Reply:
> we can very readily provide it, we're in control of this method and llvm will just ignore it

So, your plan would be to change all other calls and possibly the API of the functions calling `const_bitcast` to add the missing parameters if needed?
Or will you change the other calls so that they call another function than `const_bitcast`?

Reply:
> So, your plan would be to change all other calls and possibly the API of the functions calling `const_bitcast` to add the missing parameters if needed?
> Or will you change the other calls so that they call another function than `const_bitcast`?

ah yea we should probably just add a new function for this specific case and have llvm forward to `const_bitcast` and gcc do the checks and fallback logic

Reply:
do you mean by this comment you'd prefer to block this PR on that landing?

Reply:
@antoyo, is this PR still blocked?

Reply:
We're at the final step of running [the tests in the CI](https://github.com/rust-lang/rust/pull/146414). It should be a matter of a few weeks before it's merged.
So, if you could still wait a bit, that would be appreciated.

Reply:
If you prefer, I could also try to test this locally.

Reply:
no worries, just wasn't sure what the status was considering I was gone for a bit